File: rust/client/src/components/shared/Icon/Icon.js
import React, {memo} from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import classes from './Icon.css';

import {ReactComponent as ClockIcon} from './images/clock-solid.svg';
import {ReactComponent as EraserIcon} from './images/eraser-solid.svg';
import {ReactComponent as PencilIcon} from './images/pencil-alt-solid.svg';
import {ReactComponent as PlusIcon} from './images/plus-circle-solid.svg';
import {ReactComponent as SaveIcon} from './images/save-solid.svg';
import {ReactComponent as SmileIcon} from './images/smile-solid.svg';
import {ReactComponent as SpinnerIcon} from './images/spinner-solid.svg';
import {ReactComponent as TimesIcon} from './images/times-solid.svg';
import {ReactComponent as TrashIcon} from './images/trash-alt-solid.svg';

Icon.propTypes = {
  className: PropTypes.string,
  icon: PropTypes.string.isRequired,
  title: PropTypes.string,
  width: PropTypes.number,
  onClick: PropTypes.func
};

Icon.defaultProps = {
  width: 20
};

function Icon({className, icon, title, width, onClick}) {
  const props = {
    className,
    title,
    width,
    onClick,
    style: {verticalAlign: 'sub'}
  };

  switch (icon) {
    case 'clock':
      return <ClockIcon {...props} />;
    case 'eraser':
      return <EraserIcon {...props} />;
    case 'pencil':
      return <PencilIcon {...props} />;
    case 'plus':
      return <PlusIcon {...props} />;
    case 'save':
      return <SaveIcon {...props} />;
    case 'smile':
      return <SmileIcon {...props} />;
    case 'spinner':
      return (
        <SpinnerIcon
          {...props}
          className={cn(props.className, classes.spinner)}
        />
      );
    case 'times':
      return <TimesIcon {...props} />;
    case 'trash':
      return <TrashIcon {...props} />;

    default:
      throw Error(`Unknown icon type: '${icon}'`);
  }
}

export default memo(Icon);

File: rust+yew/client/src/components/shared/icon/icon.rs
use super::images::{
    ClockIcon, EraserIcon, IconProps, PencilIcon, PlusIcon, SaveIcon, SmileIcon, SpinnerIcon,
    TimesIcon, TrashIcon,
};
use super::IconType;
use yew::prelude::*;

#[derive(Properties, PartialEq)]
pub struct Props {
    #[prop_or_default]
    pub class: Classes,
    pub icon: IconType,
    #[prop_or_default]
    pub title: String,
    #[prop_or(20)]
    pub width: i32,
    #[prop_or_default]
    pub on_click: Callback<MouseEvent>,
}

#[function_component(Icon)]
pub fn icon(props: &Props) -> Html {
    log::trace!("render Icon");

    let icon_props = IconProps {
        class: props.class.clone(),
        style: "vertical-align: sub;".to_string(),
        title: props.title.clone(),
        width: props.width,
        on_click: props.on_click.clone(),
    };

    match props.icon {
        IconType::Clock => html! {<ClockIcon ..icon_props />},
        IconType::Eraser => html! {<EraserIcon ..icon_props />},
        IconType::Pencil => html! {<PencilIcon ..icon_props />},
        IconType::Plus => html! {<PlusIcon ..icon_props />},
        IconType::Save => html! {<SaveIcon ..icon_props />},
        IconType::Smile => html! {<SmileIcon ..icon_props />},
        IconType::Spinner => html! {<SpinnerIcon ..icon_props />},
        IconType::Times => html! {<TimesIcon ..icon_props />},
        IconType::Trash => html! {<TrashIcon ..icon_props />},
    }
}
